### PR TITLE
refactor: extract federation delete event categorizing

### PR DESCRIPTION
**Pull request by thisisamir98**
# Refactoring Federation Delete Event Categorization

## Description:
This PR introduces a refactoring of the processFederationDeleteEvent function to enhance its readability, maintainability, and testability. The primary focus of this refactoring is to extract the logic responsible for categorizing conversations affected by a federation delete event into a separate utility file, ConversationFederationUtils.ts.

## Changes Made:

- Extracted Logic: The logic responsible for categorizing conversations to leave, conversations to mark as disabled, and conversations to delete users has been extracted from the onFederationDelete function in the ConversationRepository and placed within the processFederationDeleteEvent function in the new utility file.

- Utility Functions: The utility functions within ConversationFederationUtils.ts provide a more modular and testable approach to handling the federation delete event logic.

- Unit Tests: New unit test cases have been added to ensure comprehensive test coverage of the refactored logic. These test cases cover various scenarios involving different conversation types, connection statuses, and user domains.

- Test Utilities: Utility functions for generating mock conversations and users have been used to provide realistic and controlled test scenarios.

## Impact:

- Improved Readability: The logic for handling federation delete events is now organized and encapsulated in separate utility functions, enhancing code readability and making it easier to understand.
- Enhanced Testability: The refactored code is more modular, which enables more thorough and targeted unit testing of individual components.
- Maintainability: This refactoring sets the foundation for easier maintenance and future updates, as the federation event handling logic is isolated in a separate file.

## Testing:
New unit tests have been added to the test suite to validate the correctness of the refactored processFederationDeleteEvent logic in various scenarios.

### CODE REVIEW

File: test/helper/ConversationGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import {ConversationVerificationState} from 'src/script/conversation/ConversationVerificationState';
+import {Conversation} from 'src/script/entity/Conversation';
+import {User} from 'src/script/entity/User';
+import {createUuid} from 'Util/uuid';
+
+export function generateAPIConversation(
+  id: QualifiedId = {id: createUuid(), domain: 'test.wire.link'},
+  conversationType = CONVERSATION_TYPE.REGULAR,
+  conversationProtocol = ConversationProtocol.PROTEUS,
+  overwites?: Partial<ConversationDatabaseData>,
```
Comment (atomrc) [inline test/helper/ConversationGenerator.ts]: I think we should use an object for those params, this allow settings only a few of them without doing `generateAPIConversation(undefined, undefined, PROTEUS)` for example. WDYT?

File: test/helper/ConversationGenerator.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  };
+}
+
+export function generateConversation(
+  conversationType = CONVERSATION_TYPE.REGULAR,
+  connectionStatus = ConnectionStatus.ACCEPTED,
+  conversationProtocol = ConversationProtocol.PROTEUS,
+  id?: QualifiedId,
+  users?: User[],
+  overwites?: Partial<ConversationDatabaseData>,
```
Comment (atomrc) [inline test/helper/ConversationGenerator.ts]: Same here, we could use an object

Comment (thisisamir98) [reply test/helper/ConversationGenerator.ts]: https://github.com/wireapp/wire-webapp/pull/15611/commits/f5e2de1a522ffd58294088def3d83f46618b5539

File: test/helper/ConversationGenerator.ts (hunk 1 above)
Comment (thisisamir98) [reply test/helper/ConversationGenerator.ts]: https://github.com/wireapp/wire-webapp/pull/15611/commits/f5e2de1a522ffd58294088def3d83f46618b5539

File: src/script/conversation/ConversationFederationUtils.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  conversations: Conversation[],
+): FederationDeleteResult {
+  const result: FederationDeleteResult = {
+    conversationsToLeave: [],
+    conversationsToDisable: [],
+    conversationsToDeleteUsers: [],
+  };
+
+  conversations.forEach(conversation => {
+    if (conversation.domain === deletedDomain && !conversation.is1to1()) {
```
Comment (przemvs) [inline src/script/conversation/ConversationFederationUtils.ts]: Move conversation.is1to1() to variable, and maybe another also, wdyt? 

```
const is1to1 = conversation.is1to1();
const userEntity = conversation.firstUserEntity();
const userEntities = conversation.allUserEntities();
```

**Review by przemvs — APPROVED** [review]
1 thing, LGTM!

File: src/script/conversation/ConversationFederationUtils.ts (hunk 3 above)
Comment (thisisamir98) [reply src/script/conversation/ConversationFederationUtils.ts]: https://github.com/wireapp/wire-webapp/pull/15611/commits/09b1eaa40700be01eb5550f8560bc9b60feeeccd

**Pull request merged**